pkg/csi/service/wcp/controller.go

```diff
@@ -1021,63 +1021,68 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 		}
 		if !isVADeleted {
 			log.Debugf("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
-			for k, v := range volumeAttachment.Status.AttachmentMetadata {
-				if k == common.AttributeVmUUID {
-					log.Debugf("controllerUnpublishVolumeInternal: vmuuid value: %q", v)
-					vcdcMap, err := getDatacenterFromConfig(c.manager.CnsConfig)
-					if err != nil {
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"failed to get datacenter from config with error: %+v", err)
-					}
-					var vCenterHost, dcMorefValue string
-					for key, value := range vcdcMap {
-						vCenterHost = key
-						dcMorefValue = value
-					}
-					vc, err := c.manager.VcenterManager.GetVirtualCenter(ctx, vCenterHost)
-					if err != nil {
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"cannot get virtual center %s from virtualcentermanager while attaching disk with error %+v",
-							vc.Config.Host, err)
-					}
-					// Connect to VC.
-					err = vc.Connect(ctx)
-					if err != nil {
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"failed to connect to Virtual Center: %s", vc.Config.Host)
-					}
-					podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
-					if err != nil {
-						if err == cnsvsphere.ErrVMNotFound {
-							log.Infof("virtual machine not found for vmUUID %q. "+
-								"Thus, assuming the volume is detached.", v)
-							break
+			// Check if the NodeName in the VA is same as the NodeName from the ControllerUnPublishVolume.
+			// This is needed to return a successful ControllerUnPublishVolume response when a pod is
+			// rescheduled onto another worker node(ESX host). This is a no-op for CNS-CSI in Supervisor.
+			if volumeAttachment.Spec.NodeName == req.NodeId {
```

Comment:
The diff is huge, but the only change added is this if {} condition.

Reply:
Makes sense.
Or we can also return early if the condition is false, and avoid executing the fake attach workflow as the PV might have been fake attached on some other node and we may not want to clear the pvc annotation. Probably something like:

```
if volumeAttachment.Spec.NodeName != req.NodeId {
       log.Infof("controllerUnpublishVolumeInternal: Node name in the volumeAttachment : %q is not same as "+
       "node name in the ControllerUnPublishVolume request : %q. Thus, assuming the volume is detached.",
        volumeAttachment.Spec.NodeName, req.NodeId)
	
	return &csi.ControllerUnpublishVolumeResponse{}, "", nil
	
}
```